### PR TITLE
Migrate to Scala 2.13

### PR DESCRIPTION
**Pull request by FrancoisDevemy5**
Update the build and migrate `IpSet` to the new collection framework.

There are now two versions of `IpSet`, one for 2.12 which is the same as before and one for 2.13, which is nearly the same as before. The goal is to not touch the logic at all except where we need to upgrade.
The best way to review the changes is to manually make a diff between the 2.12 and 2.13 versions of `IpSet`.

Some docs to understand the new collections:
- [the architecture of scala 2.13’s collections](https://docs.scala-lang.org/overviews/core/architecture-of-scala-213-collections.html)
- [implementing custom collections](https://docs.scala-lang.org/overviews/core/custom-collections.html)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -3,7 +3,7 @@ lazy val root = (project in file("."))
     name := "ipaddr",
     organization := "com.risksense",
     version := "1.0.2",
-    scalaVersion := "2.12.6",
+    scalaVersion := "2.13.5",
```
Comment (axuan25) [inline build.sbt]:
```suggestion
    scalaVersion := "2.13.6",
```

File: project/build.properties [hunk 2]
```diff
@@ -1 +1 @@
-sbt.version = 1.2.1
+sbt.version = 1.5.0
```
Comment (axuan25) [inline project/build.properties]:
```suggestion
sbt.version = 1.5.2
```

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+      if (networkSeq.isEmpty) {
+        throw new IpaddrException("Cannot create IpRange from an empty IpSet.")
+      } else {
+        val ip1 = IpAddress(networkSeq.head.allHosts.head)
+        val ip2 = IpAddress(networkSeq.last.allHosts.last)
+        new IpRange(ip1, ip2)
+      }
+    } else {
+      throw new IpaddrException(
+        "The input IpRange does not represent a single contiguous sequence of addresses")
```
Comment (axuan25) [inline src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:75]: I presume throwing is how the library currently behaves

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    * input IpRange are merged with other Networks of this set to form more concise blocks.
+    *
+    * @param that an IpRange object.
+    * @return an IpSet resulting from addition of `that` IpRange to `this`
+    */
+  def +(that: IpRange): IpSet = { // scalastyle:ignore method.name
+    // Create a new IpSet from input IpRange
+    val newSet = IpSet(that)
+    // Perform union with this IpSet
+    val unionResult = this | newSet
```
Comment (axuan25) [inline src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:185]: 👍 !

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+     */
+    @tailrec
+    def intersectRecurse(
+        s1: Seq[IpNetwork],
+        s2: Seq[IpNetwork],
+        result: Seq[IpNetwork]): Seq[IpNetwork] = {
+      if (s1.isEmpty || s2.isEmpty) {
+        result
+      } else {
+        if (s1.head == s2.head) {
```
Comment (axuan25) [inline src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:379]: Anyway to remove calls to `.head` ?
```suggestion
        if (s1.headOption == s2.headOption) {
```
Should also work no?

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala (hunk 3 above)
Comment (FrancoisDevemy5) [reply src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:75]: Yes

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala (hunk 5 above)
Comment (FrancoisDevemy5) [reply src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:379]: It is not the goal of this PR to change behaviour like this.

Comment (axuan25) [reply src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:379]: OK, so I take it that this was already coded this way (didn't check 😶 ) 👍

File: src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+  /** Add a IpNetwork to this IpSet
+    *
+    * Adds a [[IpNetwork]] to this IpSet and returns a new IpSet. Where possible the input Network
+    * is merged with other Networks of `this` set to form more concise blocks.
+    *
+    * @param that a IpNetwork object.
+    * @return an IpSet resulting from addition of `that` IpNetwork to `this`
+    */
+  override def +(that: IpNetwork): IpSet = { // scalastyle:ignore method.name
```
Comment (axuan25) [inline src/main/scala-2.13+/com/risksense/ipaddr/IpSet.scala:164]: 👍  previous implem isn't forthcoming about overrides

**Review by axuan25 — APPROVED** [review]
Re-brushed up on Scala 2.13 collections, checked previous implem (this time 😄 ), LGTM !

File: project/build.properties [hunk 7]
```diff
@@ -1 +1 @@
-sbt.version = 1.2.1
+sbt.version = 1.5.2
```
Comment (axuan25) [inline project/build.properties]:
```suggestion
sbt.version = 1.5.3
```
😅

**Pull request merged**